HPC/WRF.md

```diff
@@ -0,0 +1,267 @@
+## WRF
+
+The Weather Research and Forecasting (WRF) model is one of the most used numerical weather prediction (NWP) system. WRF is used extensively for research and real-time forecasting with over 52,570 registered users in over 184 countries throughout the world as of Dec 2020. Large amount of computation resources are required for each simulation, especially for high resolution simulations. Compute and MPI performance are very important for WRF performance.
+
+## WRF correctness fix
+We have observed small numeric differences between WRF output models (v4.2.2) generated on Arm64 and intel platforms. The issue is caused by rounding error at last position (1ulp) from transcendental functions and several compiler optimization options. After identifying the root cause, we implemented [a patch](https://github.com/wrf-model/WRF/pull/1773) that allows bit-by-bit matching results on Graviton and intel processors for Ubuntu 18.04 with GCC 10.2 compiler. This patch works for both v4.2.2 and v4.5 (latest version). 
+
+## Build and run WRF with GCC on arm64
```

Comment:
having to build the complier will care ppl away, can we just assume al2023 or ubuntu 22.04

Reply:
For OpenFOAM GCC compiler v9.4 and v12.0 had similar perf. So, we can just say 9.4+ should be good enough unless you have data to say 10.2 is better for WRF. Just helps avoid user the need to get another compiler over what is provided in PC.

Reply:
I tried the build on default gcc/gfortran 9.4 and openmpi; pnetcdf failed at configure.

```
checking how to get verbose linking output from /opt/amazon/openmpi/bin//mpif90... configure: WARNING: compilation failed

checking for Fortran 77 libraries of /opt/amazon/openmpi/bin//mpif90... 
checking whether /opt/amazon/openmpi/bin//mpif90 is a valid MPI compiler... no
configure: error: 
   -----------------------------------------------------------------------
     Invalid MPI Fortran 77 compiler: "/opt/amazon/openmpi/bin//mpif90"
     A working MPI Fortran 77 compiler is required. Please specify the
     location of a valid MPI Fortran 77 compiler, either in the MPIF77
     environment variable or through --with-mpi configure flag. Abort.
   -----------------------------------------------------------------------
make: *** No targets specified and no makefile found.  Stop.

ubuntu@ip-10-0-1-52: mpif90 --version
GNU Fortran (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Reply:
Didn't quite understand this comment. I found mpif77 here:

$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"

$ which mpif77
/opt/amazon/openmpi/bin//mpif77

Reply:
I have tried MPIF77=mpif77 as suggested by config.log.  It also fails to configure.

Reply:
> I can remove building compilers from the instruction. But parallel cluster supports only ubuntu 1804 and 2004, redhat 7 and al2 at the moment. The default gcc is either 7.5.0 or 9.4.0. I think newer compiler (GCC-10.2 or later) does help with the performance for WRF and OpenFoam.

Still we don't need to build the compiler. we can upgrade 9.x to 10.x on ubuntu 20.04.
did we check the parallel cluster roadmap for newer OS support? should we make the request?

Reply:
We need to make the request for the compiler update if we want the default compiler to work.

Reply:
Its easier to tell people how to get the necessary compiler from alternate package sources.  There are PPAs on Ubuntu for this, having to build GCC from source is a large source of friction.

Reply:
Thanks, I have removed building GCC from source from this guide.

Reply:
I have added parallel cluster configuration to automatically upgrade to gcc-11 by a post-configure script.